### PR TITLE
Converting 'Discover' to Jetpack Compose

### PR DESCRIPTION
**Pull request by chrisbanes**

### CODE REVIEW

File: common-ui-compose/src/main/java/app/tivi/common/compose/Carousel.kt [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import androidx.compose.ui.unit.Dp
+import androidx.compose.ui.unit.dp
+
+@Composable
+fun <T> Carousel(
+    items: List<T>,
+    modifier: Modifier = Modifier,
+    contentPadding: InnerPadding = InnerPadding(0.dp),
+    itemSpacing: Dp = 0.dp,
+    verticalGravity: Alignment.Vertical = Alignment.Top,
```
Comment (nickbutcher) [inline common-ui-compose/src/main/java/app/tivi/common/compose/Carousel.kt:34]: I'd rename `verticalAlignment` in line with upcoming changes.

File: common-ui-compose/src/main/java/app/tivi/common/compose/PosterCard.kt [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    show: TiviShow,
+    poster: TmdbImageEntity? = null,
+    onClick: (() -> Unit)? = null,
+    modifier: Modifier = Modifier
+) {
+    Card(modifier = modifier) {
+        Stack(
+            modifier = if (onClick != null) Modifier.clickable(onClick = onClick) else Modifier
+        ) {
+            ProvideEmphasis(EmphasisAmbient.current.medium) {
```
Comment (nickbutcher) [inline common-ui-compose/src/main/java/app/tivi/common/compose/PosterCard.kt:46]: So if the poster is faded in, this is on top of the text? If so can we omit the text when it's loaded? Does Coil expose info to achieve this?

File: ui-discover/src/main/java/app/tivi/home/discover/Discover.kt [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+import app.tivi.util.TiviDateFormatter
+import com.google.android.material.composethemeadapter.MdcTheme
+import dev.chrisbanes.accompanist.coil.CoilImage
+
+internal fun ViewGroup.composeDiscover(
+    state: LiveData<DiscoverViewState>,
+    actioner: (DiscoverAction) -> Unit,
+    tiviDateFormatter: TiviDateFormatter,
+    textCreator: DiscoverTextCreator
+): Any = setContent(Recomposer.current()) {
```
Comment (nickbutcher) [inline ui-discover/src/main/java/app/tivi/home/discover/Discover.kt:91]: Use `ComposeView` instead?

File: ui-discover/src/main/java/app/tivi/home/discover/Discover.kt [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+    Providers(
+        TiviDateFormatterAmbient provides tiviDateFormatter,
+        DiscoverTextCreatorAmbient provides textCreator
+    ) {
+        MdcTheme {
+            LogCompositions("MdcTheme")
```
Comment (nickbutcher) [inline ui-discover/src/main/java/app/tivi/home/discover/Discover.kt:97]: Remove?

File: ui-discover/src/main/java/app/tivi/home/discover/Discover.kt [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+        Stack(Modifier.fillMaxSize()) {
+            var appBarHeight by rememberMutableState { 0 }
+
+            LazyColumn(Modifier.fillMaxSize()) {
+                item {
+                    val height = with(DensityAmbient.current) { appBarHeight.toDp() } + 16.dp
+                    Spacer(Modifier.preferredHeight(height))
+                }
+
+                state.nextEpisodeWithShowToWatched?.also { nextEpisodeToWatch ->
```
Comment (nickbutcher) [inline ui-discover/src/main/java/app/tivi/home/discover/Discover.kt:127]: nit: `also` seems strange here—wold expect `?.let` for a null check

File: ui-discover/src/main/java/app/tivi/home/discover/Discover.kt [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+        Row(Modifier.padding(horizontal = 16.dp, vertical = 8.dp)) {
+            if (poster != null) {
+                PosterCard(
+                    show = show,
+                    poster = poster,
+                    modifier = Modifier.preferredWidth(64.dp).aspectRatio(2 / 3f)
+                )
+            }
+
+            Spacer(Modifier.preferredWidth(16.dp))
```
Comment (nickbutcher) [inline ui-discover/src/main/java/app/tivi/home/discover/Discover.kt:216]: Should this go inside the above `if` i.e. don't you only want this is there is a poster?

File: ui-discover/src/main/java/app/tivi/home/discover/Discover.kt [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+    }
+}
+
+@OptIn(ExperimentalAnimationApi::class)
+@Composable
+private fun Header(
+    title: String,
+    loading: Boolean = false,
+    action: (@Composable () -> Unit)? = null,
+    modifier: Modifier = Modifier
```
Comment (nickbutcher) [inline ui-discover/src/main/java/app/tivi/home/discover/Discover.kt:311]: FIrst optional param

File: ui-discover/src/main/java/app/tivi/home/discover/Discover.kt [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+        )
+    }
+}
+
+@OptIn(ExperimentalAnimationApi::class)
+@Composable
+private fun Header(
+    title: String,
+    loading: Boolean = false,
+    action: (@Composable () -> Unit)? = null,
```
Comment (nickbutcher) [inline ui-discover/src/main/java/app/tivi/home/discover/Discover.kt:310]: Make last param to enable trailing lambda.

File: ui-discover/src/main/java/app/tivi/home/discover/Discover.kt [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+private fun DiscoverAppBar(
+    loggedIn: Boolean,
+    user: TraktUser?,
+    refreshing: Boolean,
+    onRefreshActionClick: () -> Unit,
+    onUserActionClick: () -> Unit,
+    modifier: Modifier = Modifier
+) {
+    Surface(
+        color = MaterialTheme.colors.surface.copy(alpha = 0.93f),
```
Comment (nickbutcher) [inline ui-discover/src/main/java/app/tivi/home/discover/Discover.kt:350]: Factor the alpha out somewhere?

File: ui-discover/src/main/java/app/tivi/home/discover/Discover.kt [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+                        loggedIn -> IconResource(R.drawable.ic_person)
+                        else -> IconResource(R.drawable.ic_person_outline)
+                    }
+                }
+            }
+        }
+    }
+}
+
+@Preview
```
Comment (nickbutcher) [inline ui-discover/src/main/java/app/tivi/home/discover/Discover.kt:403]: No theme on previews?

File: ui-discover/src/main/java/app/tivi/home/discover/DiscoverFragment.kt [hunk 11]
```diff
@@ -18,173 +18,75 @@ package app.tivi.home.discover
...
-            addItemDecoration(SpacingItemDecorator(paddingLeft))
-        }
-
-        binding.followedAppBar.doOnSizeChange {
-            binding.summaryRv.updatePadding(top = it.height)
-            binding.summarySwipeRefresh.setProgressViewOffset(
-                true, 0,
-                it.height + binding.summarySwipeRefresh.progressCircleDiameter / 2
+    ): View? {
+        return FrameLayout(requireContext()).apply {
```
Comment (nickbutcher) [inline ui-discover/src/main/java/app/tivi/home/discover/DiscoverFragment.kt:52]: ComposeView

File: ui-discover/src/main/java/app/tivi/home/discover/DiscoverTextCreator.kt [hunk 12]
```diff
@@ -17,12 +17,15 @@
...
 
 import android.content.Context
+import androidx.compose.runtime.staticAmbientOf
 import app.tivi.data.entities.Episode
 import app.tivi.data.entities.Season
 import dagger.hilt.android.qualifiers.ActivityContext
 import javax.inject.Inject
 
-internal class DiscoverTextCreator @Inject constructor(
+val DiscoverTextCreatorAmbient = staticAmbientOf<DiscoverTextCreator>()
```
Comment (nickbutcher) [inline ui-discover/src/main/java/app/tivi/home/discover/DiscoverTextCreator.kt:26]: Ambient feels a bit much for this. Could just make an extension method on `Context`?

File: ui-discover/src/main/java/app/tivi/home/discover/Discover.kt [hunk 13]
```diff
@@ -0,0 +1,424 @@
...
+                        Header(title = stringResource(R.string.discover_keep_watching_title))
+                    }
+                    item {
+                        NextEpisodeToWatch(
+                            show = nextEpisodeToWatch.show,
+                            poster = nextEpisodeToWatch.poster,
+                            season = nextEpisodeToWatch.season,
+                            episode = nextEpisodeToWatch.episode,
+                            modifier = Modifier.fillMaxWidth().clickable {
+                                actioner(
```
Comment (nickbutcher) [inline ui-discover/src/main/java/app/tivi/home/discover/Discover.kt:138]: This looks odd?! An onClick defining UI?

File: common-ui-compose/src/main/java/app/tivi/common/compose/Carousel.kt (hunk 1 above)
Comment (chrisbanes) [reply common-ui-compose/src/main/java/app/tivi/common/compose/Carousel.kt:34]: I'll wait for now, still on alpha01 atm

File: ui-discover/src/main/java/app/tivi/home/discover/DiscoverFragment.kt (hunk 11 above)
Comment (chrisbanes) [reply ui-discover/src/main/java/app/tivi/home/discover/DiscoverFragment.kt:52]: Done.

File: ui-discover/src/main/java/app/tivi/home/discover/Discover.kt (hunk 6 above)
Comment (chrisbanes) [reply ui-discover/src/main/java/app/tivi/home/discover/Discover.kt:180]: Good catch

File: ui-discover/src/main/java/app/tivi/home/discover/Discover.kt (hunk 8 above)
Comment (chrisbanes) [reply ui-discover/src/main/java/app/tivi/home/discover/Discover.kt:274]: Done

File: ui-discover/src/main/java/app/tivi/home/discover/Discover.kt (hunk 13 above)
Comment (chrisbanes) [reply ui-discover/src/main/java/app/tivi/home/discover/Discover.kt:102]: Yeaaahhhhh, something I generally need to fix in Tivi. Basically I'm missing the 'processor' / 'actor' bit of MVI currently.

File: ui-discover/src/main/java/app/tivi/home/discover/Discover.kt (hunk 10 above)
Comment (chrisbanes) [reply ui-discover/src/main/java/app/tivi/home/discover/Discover.kt:367]: Pssstttt, I don't use Previews 🙈

File: ui-discover/src/main/java/app/tivi/home/discover/Discover.kt (hunk 4 above)
Comment (chrisbanes) [reply ui-discover/src/main/java/app/tivi/home/discover/Discover.kt]: I like to keep these. They get removed anyway in the release build.

**Pull request merged**